From e0b4c12cfd3a7c2b3ae5ccc17a6335d9a17f0ea8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?C=C3=A9dric=20Boutillier?= <boutil@debian.org>
Date: Sat, 3 Mar 2018 14:03:54 +0100
Subject: [PATCH] Deprecate SHA1 and check availibility of the strongs
 SHA256/SHA512

- replace SHA1 with SHA256
- if none of the SHA256/SHA512 is listed in checksums.yaml, quit with an
  error

https://salsa.debian.org/ruby-team/gem2deb/-/commit/e0b4c12cfd3a7c2b3ae5ccc17a6335d9a17f0ea8
https://bugs.launchpad.net/ubuntu/+source/gem2deb/+bug/1753601
---
 usr/lib/ruby/vendor_ruby/gem2deb/gem2tgz.rb | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

diff --git a/usr/lib/ruby/vendor_ruby/gem2deb/gem2tgz.rb b/usr/lib/ruby/vendor_ruby/gem2deb/gem2tgz.rb
index 456315e..a201e8a 100644
--- a/usr/lib/ruby/vendor_ruby/gem2deb/gem2tgz.rb
+++ b/usr/lib/ruby/vendor_ruby/gem2deb/gem2tgz.rb
@@ -125,13 +125,23 @@ module Gem2Deb
 
     def verify_and_strip_checksums
       checksums = read_checksums
-      [Digest::SHA1, Digest::SHA512].each do |digest|
+      digests=[Digest::SHA256, Digest::SHA512]
+      unverified_checksums = []
+      digests.each do |digest|
         hash_name = digest.name.sub(/^Digest::/,'')
-        ["data.tar.gz", "metadata.gz"].each do |f|
-          unless correct_checksum?(digest, f, checksums[hash_name][f])
-            puts "E: (#{gem}) the #{hash_name} checksum for #{f} is inconsistent with the one recorded in checksums.yaml.gz"
-            exit(1)
+        if checksums.has_key? hash_name
+          ["data.tar.gz", "metadata.gz"].each do |f|
+            unless correct_checksum?(digest, f, checksums[hash_name][f])
+              puts "E: (#{gem}) the #{hash_name} checksum for #{f} is inconsistent with the one recorded in checksums.yaml.gz"
+              exit(1)
+            end
           end
+        else
+          unverified_checksums << hash_name
+        end
+        if unverified_checksums.size == digests.size
+          puts "E: (#{gem}) missing checksums for #{unverified_checksums.join(", ")}. Only #{checksums.keys.join(", ")} recorded in checksums.yaml.gz"
+          exit(1)
         end
       end
       FileUtils.rm_f('checksums.yaml.gz')
-- 
GitLab

